### PR TITLE
Fix #582 by extending support to Clocktime and supporting null values when comparing with a Date or Clocktime

### LINKED ISSUE #582: (Function) Extend support of Min and Max to Clocktimes & support null values
Which ought to be comparable with dates (a clocktime directly from the DB might be a Date)
Also if one of the two values is null while the other is a Date or Clocktime, comparison should end up with the non-null value returned (it currently crashes).

### PR DESCRIPTION
**Pull request by kharybdys**
Note that this does introduce a dependency from Navajo functions to Navajo document (as I'm now referring Clocktime) but I think this is acceptable

### CODE REVIEW

File: core/com.dexels.navajo.function/src/com/dexels/navajo/functions/Max.java [hunk 1]
```diff
@@ -64,17 +64,24 @@ public final Object evaluate() throws TMLExpressionException {
             throw new TMLExpressionException("Invalid number of arguments for Max()");
     }
     
-    private Object max(Object a, Object b)
+    private Object max( Object a, Object b )
     {
-        if (a instanceof Date)
+        if ( a != null && ( a instanceof Date || a instanceof ClockTime ) )
```
Comment (roelofkemp) [inline core/com.dexels.navajo.function/src/com/dexels/navajo/functions/Max.java:69]: you can leave out the null check, see: https://stackoverflow.com/questions/2950319/is-null-check-needed-before-calling-instanceof

File: core/com.dexels.navajo.function/src/com/dexels/navajo/functions/Max.java [hunk 2]
```diff
@@ -86,20 +93,27 @@ private Object max(Object a, Object b)
         }
         else
         {
-            if ((b instanceof Date))
+            if ( b != null && ( b instanceof Date || b instanceof ClockTime ) )
```
Comment (roelofkemp) [inline core/com.dexels.navajo.function/src/com/dexels/navajo/functions/Max.java:96]: you can leave out the null check, see: https://stackoverflow.com/questions/2950319/is-null-check-needed-before-calling-instanceof

File: core/com.dexels.navajo.function/src/com/dexels/navajo/functions/Max.java [hunk 3]
```diff
@@ -86,20 +93,27 @@ private Object max(Object a, Object b)
...
+                double value2 = Utils.getDoubleValue( b );
+                double max = ( value1 > value2 ) ? value1 : value2;
 
-                if (a instanceof Double || b instanceof Double)
-                    return Double.valueOf(max);
+                if ( a instanceof Double || b instanceof Double )
+                    return Double.valueOf( max );
                 else
-                    return Integer.valueOf((int) max);
+                    return Integer.valueOf( (int) max );
```
Comment (roelofkemp) [inline core/com.dexels.navajo.function/src/com/dexels/navajo/functions/Max.java:116]: Please don't mix features with code formatting

File: core/com.dexels.navajo.function/src/com/dexels/navajo/functions/Min.java [hunk 4]
```diff
@@ -66,17 +64,24 @@ public final Object evaluate() throws TMLExpressionException {
         } else
             throw new TMLExpressionException("Invalid number of arguments for Max()");
     }
-    private Object min(Object a, Object b)
+
+    private Object min( Object a, Object b )
     {
-        if (a instanceof Date)
+        if ( a != null && ( a instanceof Date || a instanceof ClockTime ) )
```
Comment (roelofkemp) [inline core/com.dexels.navajo.function/src/com/dexels/navajo/functions/Min.java:70]: you can leave out the null check, see: https://stackoverflow.com/questions/2950319/is-null-check-needed-before-calling-instanceof

File: core/com.dexels.navajo.function/src/com/dexels/navajo/functions/Min.java [hunk 5]
```diff
@@ -88,9 +93,16 @@ private Object min(Object a, Object b)
         }
         else
         {
-            if ((b instanceof Date))
+            if ( b != null && ( b instanceof Date || b instanceof ClockTime ) )
```
Comment (roelofkemp) [inline core/com.dexels.navajo.function/src/com/dexels/navajo/functions/Min.java:96]: you can leave out the null check, see: https://stackoverflow.com/questions/2950319/is-null-check-needed-before-calling-instanceof

**Review by roelofkemp — APPROVED** [review]

File: core/com.dexels.navajo.function/src/com/dexels/navajo/functions/Max.java (hunk 3 above)
Comment (kharybdys) [reply core/com.dexels.navajo.function/src/com/dexels/navajo/functions/Max.java:116]: I don't agree. When new code is formatted a different way than existing code, the rest should be adjusted to match. Maybe it should've been in a separate commit so you can ignore the formatting commit, but every formatting change has a chance to introduce errors so should not be ignored at a review anyways.

Whether it's "correct" to format it my way is another discussion that we can have ;)

File: core/com.dexels.navajo.function/src/com/dexels/navajo/functions/Max.java (hunk 1 above)
Comment (kharybdys) [reply core/com.dexels.navajo.function/src/com/dexels/navajo/functions/Max.java:69]: The null check is not there for the instanceof. Right now if you enter a Date and a null value you end up at the coded exception "Cannot compare a date with a non-date". But it should be possible to compare these with the Date always winning, both at Min and Max.

File: core/com.dexels.navajo.function/src/com/dexels/navajo/functions/Max.java (hunk 2 above)
Comment (kharybdys) [reply core/com.dexels.navajo.function/src/com/dexels/navajo/functions/Max.java:96]: The null check is not there for the instanceof. Right now if you enter a Date and a null value you end up at the coded exception "Cannot compare a date with a non-date". But it should be possible to compare these with the Date always winning, both at Min and Max.

File: core/com.dexels.navajo.function/src/com/dexels/navajo/functions/Min.java (hunk 4 above)
Comment (kharybdys) [reply core/com.dexels.navajo.function/src/com/dexels/navajo/functions/Min.java:70]: The null check is not there for the instanceof. Right now if you enter a Date and a null value you end up at the coded exception "Cannot compare a date with a non-date". But it should be possible to compare these with the Date always winning, both at Min and Max.

File: core/com.dexels.navajo.function/src/com/dexels/navajo/functions/Min.java (hunk 5 above)
Comment (kharybdys) [reply core/com.dexels.navajo.function/src/com/dexels/navajo/functions/Min.java:96]: The null check is not there for the instanceof. Right now if you enter a Date and a null value you end up at the coded exception "Cannot compare a date with a non-date". But it should be possible to compare these with the Date always winning, both at Min and Max.

**Review by gballintijn — APPROVED** [review]

**Comment by gballintijn** [comment]
FYI: Not a fan of the amount and placements of spaces and empty (non-code) lines. I think it degrades readability.

**Comment by kharybdys** [comment]
And interestingly I think it improves readability. Am I getting old?

We should have a default (company-wide) way of formatting things though, at least up to some level, but that doesn't exist (yet) :(

**Pull request merged**